ignite/pkg/protoanalysis/protoutil/utils.go

```diff
@@ -0,0 +1,286 @@
+		if _, ok := c.Node().(*proto.Syntax); ok {
+			c.InsertAfter(v)
+			inserted = true
+			return false
+		}
+		// Continue until we insert.
+		return true
+	})
+	if inserted {
+		return nil
+	}
+	return errors.New("could not find syntax statement")
+}
+
+// AddAfterPackage tries to add the given Visitee after the 'package' statement.
+//
+// If no package statement is found, returns an error.
+func AddAfterPackage(f *proto.Proto, v proto.Visitee) error {
+	inserted := false
+	Apply(f, nil, func(c *Cursor) bool {
+		if _, ok := c.Node().(*proto.Package); ok {
+			c.InsertAfter(v)
+			inserted = true
+			return false
+		}
+		// Continue until we insert.
+		return true
+	})
+	if inserted {
+		return nil
+	}
+	return errors.New("could not find package statement")
+}
+
+// Fallback logic, try and use import after a package and if that fails
+// attempts to use it after a syntax statement.
+// If that fails, returns an error.
+func importFallback(f *proto.Proto, imp *proto.Import) error {
+	if err := AddAfterPackage(f, imp); err != nil {
+		if err = AddAfterSyntax(f, imp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AddImport tries its best to add the given import *after* any other imports
+// in the file.
+//
+// If fallback is supplied, attempts to add it after the 'package'
+// statement and then the 'syntax' statement are made.
+//
+// If non of the attempts are successful, returns an error.
+func AddImports(f *proto.Proto, fallback bool, imports ...*proto.Import) (err error) {
+	// No effect.
+	if len(imports) == 0 {
+		return nil
+	}
+	importMap, inserted := make(map[string]*proto.Import), false
+	for _, i := range imports {
+		importMap[i.Filename] = i
+	}
+
+	Apply(f, nil, func(c *Cursor) bool {
+		if i, ok := c.Node().(*proto.Import); ok {
+			delete(importMap, i.Filename)
+			if next, ok := c.Next(); ok {
+				if _, ok := next.(*proto.Import); ok {
+					return true
+				}
+				for _, imp := range importMap {
+					c.InsertAfter(imp)
+				}
+				inserted = true
+				return false
+			}
+			// We're at the end (no Next())
+			for _, imp := range importMap {
+				c.InsertAfter(imp)
+			}
+			inserted = true
+			return false
+		}
+		return true
+	})
+	// return if inserted.
+	if inserted {
+		return nil
+	}
+	// else fallback if defined.
+	if fallback {
+		// if the number of improts is > 1, we can try and insert the first after
+		// the package/syntax and then recurse into AddImport with the rest (which we'll)
+		// know that we can insert after an import since we just added it.
+		imports = []*proto.Import{}
+		for _, imp := range importMap {
+			imports = append(imports, imp)
+		}
+		if len(imports) == 0 {
+			return nil
+		}
+		if err := importFallback(f, imports[0]); err != nil {
+			return err
+		}
+		// recurse with the rest. (might be empty)
+		return AddImports(f, false, imports[1:]...)
+	}
+	return errors.New("unable to add import, no import statements found")
+}
+
+// TODO: Get(Least/Min)UniqueID?
```

Comment:
Same as the above comment for TODO

If there is a need for future refactoring, we can document it